### DISCUSSION TITLE
Update, Delete on DOM only after hard refresh

### DISCUSSION THREAD

**Discussion by danicrosby** [nss-evening-cohort-14/evening-client-side#287] [Capstones!] [answered]
Dani Crosby

Note: I have all my edit and delete logic on my 'Single Dream Card', not the general Dream Card. 

I am able to view Single Dream with an edit and delete button.
I can click an edit button and an edit form shows up, which I am able to update.
The new info updates in the background (firebase), but will only update on the DOM if I hard refresh.

Similarly, the delete button deletes the card off firebase and pops me back into the Dream Card View, but does not delete off the DOM in Dream Collection View unless hard refresh. Note: delete and edit *were* working when these buttons were on the Dream Card View

Below code is placed in Single Card View helped redirect from Single Card View back to Dream Card View. I'm assuming there is another (( .then )) call to fully remove from Dream Card View or a useEffect, but I am stuck on what to add and where to add it.

     case 'delete':
        deleteDream(dream.firebaseKey)
          .then(() => {
            history.push('/dreams');
          });

No errors present.
I have asked around, but most people have the same issue.

No API's, only hardcoded data.

Dream Card: 
https://github.com/danicrosby/LUCID-Front-End-Capstone/blob/main/src/components/DreamCard.js

Single Card: 
https://github.com/danicrosby/LUCID-Front-End-Capstone/blob/main/src/components/SingleDreamCard.js

Dream Form: 
https://github.com/danicrosby/LUCID-Front-End-Capstone/blob/main/src/components/DreamForm.js

Full Repo: 
https://github.com/danicrosby/LUCID-Front-End-Capstone

I have researched these articles: 
https://www.pluralsight.com/guides/how-to-unmount-react.js-node
https://dev.to/chilupa/remove-element-from-dom-in-react-way-n2l
https://stackoverflow.com/questions/55324088/why-do-i-have-to-refresh-page-manually-after-delete

Referenced previous tickets:
#284
#277
#218
https://github.com/nss-evening-cohort-14/evening-client-side/discussions/183

**Comment by TrinityChristiana** [comment, marked as answer]
@danicrosby Will add their solution.

**Reply by danicrosby** [reply]
Dr. T, Trinity, and I walked through and created a user flow chart and realized my single dreams weren't talking to my dream cards AT ALL, they weren't even connected which is why the information wasn't passed around. I also had a mixture of 'dreams' vs 'dream', so keeping everything consistent is key.

For delete we added Single State and history.push: 
<img width="469" alt="Screen Shot 2021-06-12 at 12 28 13 PM" src="https://user-images.githubusercontent.com/68397076/121784456-aa164680-cb79-11eb-98fe-de2fce0c6d44.png">

We also updated the form to include the new state/function: 
<img width="380" alt="Screen Shot 2021-06-12 at 12 29 22 PM" src="https://user-images.githubusercontent.com/68397076/121784506-d4680400-cb79-11eb-9098-f987b907c2e4.png">

For edit: After lots of digging around, the ultimate issue was I was returning ALL of my dream cards rather than just the single dream I needed to edit the form: 

Data from this: 
<img width="580" alt="Screen Shot 2021-06-12 at 12 26 32 PM" src="https://user-images.githubusercontent.com/68397076/121784400-6e7b7c80-cb79-11eb-8f8c-e75737469218.png">

To this: 
<img width="549" alt="Screen Shot 2021-06-12 at 12 33 14 PM" src="https://user-images.githubusercontent.com/68397076/121784595-5d7f3b00-cb7a-11eb-8540-daa811d2d2cc.png">

And added this useEffect and wrapped the SingleDream in the SingleView:
<img width="516" alt="Screen Shot 2021-06-12 at 12 25 28 PM" src="https://user-images.githubusercontent.com/68397076/121784362-4724af80-cb79-11eb-8144-df660054cb44.png">

**Reply by drteresavasquez** [reply]
@danicrosby this is one of the most beautiful explanations I have ever seen. You understand what happened and what you fixed and I'm in awe of your awesome! Excellent job!
